lib/runner.go

```diff
@@ -194,84 +240,98 @@ func (r *Runner) runAllTests(testFiles []string, testFolder string) {
+		buf := NewConcurrentBuffer()
+		command.Stdout = buf
+		command.Stderr = buf
+		testResult.Output = buf
+	}
+
+	// Propagate timeout information from brain to script, via the environment of the script.
 	env := os.Environ()
-	env = append(env, fmt.Sprintf("TESTBRAIN_TIMEOUT=%d", int(r.Timeout.Seconds())))
+	env = append(env, fmt.Sprintf("TESTBRAIN_TIMEOUT=%v", r.options.timeout.Seconds()))
 	command.Env = env
 
 	err := command.Start()
 	if err != nil {
-		return ErrorTestResult(testFile, err)
-	}
-	var timeoutLock sync.Mutex
-	timeoutReached := false
-	timer := time.AfterFunc(r.Timeout, func() {
-		command.Process.Kill()
-		timeoutLock.Lock()
-		defer timeoutLock.Unlock()
-		timeoutReached = true
-	})
-	err = command.Wait()
-	timer.Stop()
-	timeoutLock.Lock()
-	if timeoutReached {
-		commandOutput.WriteString(fmt.Sprintf("Killed by testbrain: Timed out after %v", r.Timeout))
+		fmt.Fprintf(r.stderr, "Test failed: %v", err)
+		testResult.Success = false
+		testResult.ExitCode = -1
+		return testResult
 	}
-	timeoutLock.Unlock()
 
-	exitCode, err := r.getErrorCode(err, command)
-	if err != nil {
-		return ErrorTestResult(testFile, err)
-	}
+	done := make(chan error)
+	go func() {
+		done <- command.Wait()
+	}()
 
-	return TestResult{
-		TestFile: testFile,
-		Success:  command.ProcessState.Success(),
-		ExitCode: exitCode,
-		Output:   string(commandOutput.Bytes()),
+	timeout := time.After(r.options.timeout)
+
+	select {
+	case <-timeout:
+		command.Process.Kill()
+		fmt.Fprintf(r.stderr, "Killed by testbrain: Timed out after %v\n", r.options.timeout)
+		testResult.Success = false
+		testResult.ExitCode = -1
+	case err = <-done:
+		testResult.ExitCode, err = r.getErrorCode(err, command)
+		if err != nil {
+			fmt.Fprintf(r.stderr, "Test failed: %v", err)
+			testResult.Success = false
+			testResult.ExitCode = -1
+			return testResult
+		}
+		testResult.Success = command.ProcessState.Success()
 	}
+
+	return testResult
 }
 
 func (r *Runner) getErrorCode(err error, command *exec.Cmd) (int, error) {
 	if command.ProcessState.Success() {
 		// Not exactly necessary, since we can check Success(),
-		// but more correct than saying status code is unknown
+		// but more correct than saying status code is unknown.
 		return 0, nil
 	}
 	if err != nil {
 		exitErr, ok := err.(*exec.ExitError)
 		if ok {
-			// The program has exited with an ExitError, we can get the error code from WaitStatus
-			// See https://golang.org/pkg/os/#ProcessState.Sys
-			// Although the docs mention syscall.WaitStatus works on Unix, it seems to work on Windows too
 			status, ok := exitErr.Sys().(syscall.WaitStatus)
 			if ok {
 				return status.ExitStatus(), nil
 			}
 		}
 
-		// There is an error but it's not an ExitError
-		// Something other than the test script failed, bubble up the error
-		return UnknownExitCode, err
+		// There is an error but it's not an ExitError.
+		// Something other than the test script failed, bubble up the error.
+		return -1, err
```

Comment:
I would recommend putting this in a var, it's not clear what the `-1` is at the first glance.